src/bluesky/simulators.py

```diff
@@ -104,3 +106,279 @@ async def check_limits_async(plan):
             else:
                 warn(f"{obj.name} has no check_value() method" f" to check if {msg.args[0]} is within its limits.")  # noqa: B028
                 ignore.append(obj)
+
+
+class RunEngineSimulator:
+    """Helps test a Bluesky plan by recording bluesky messages and
+     injecting responses according to the bluesky Message Protocol.
+
+    See bluesky docs for details of the message protocol.
+
+    Basic usage consists of
+    1) Registering various handlers to respond to anticipated messages in the experiment plan and fire
+     any needed callbacks.
+    2) Calling simulate_plan()
+    3) Examining the returned message list and making asserts against them.
+
+    Attributes
+    ----------
+    return_value
+        The return value of the most recently executed plan.
+
+    Examples
+    --------
+
+    >>> def my_plan():
+    ...     pass
+
+    >>> sim = RunEngineSimulator()
+    >>> messages = sim.simulate_plan(my_plan())
+    >>>
+    """
+
+    GROUP_ANY = "any"
+
+    def __init__(self):
+        self.message_handlers = []
+        self.callbacks = {}
+        self.next_callback_token = 0
+        self.return_value: Any
+
+    def add_handler_for_callback_subscribes(self):
+        """Add a handler that registers all the callbacks from subscribe messages so we can call them later.
+
+        You will need to call this as one of the first things if you wish to fire callbacks from the simulator.
+        """
+        self.message_handlers.append(
+            _MessageHandler(
+                lambda msg: msg.command == "subscribe",
+                lambda msg: self._add_callback(msg.args),
+            )
+        )
+
+    def add_handler(
+        self,
+        commands: Union[str, Sequence[str]],
+        handler: Callable[[Msg], object],
+        msg_filter: Optional[Union[str, Callable[[Msg], bool]]] = None,
+        index: int = 0,
+    ):
+        """Add the specified handler for a particular message.
+
+        Parameters
+        ----------
+        commands
+            The command name for the message as defined in bluesky Message Protocol, or a sequence if
+            more than one matches.
+        handler
+            A lambda that accepts a Msg and returns an object; the object is sent to the current yield
+            statement in the generator, and is used when reading values from devices, the structure of the
+            object depends on device hinting.
+        msg_filter
+            This can either be a string corresponding to the Msg obj.name property to match, or it can
+            be a predicate on the Msg which returns a bool. The default is None which will match all
+            messages.
+        index
+            An optional integer indicating where to insert the handler, the default is 0 which
+            prepends it so that newer handlers override older ones. Specify -1 to append the handler.
```

Comment:
Any negative number puts it at the end.  Should `-3` put it 3 in from the back?

Reply:
I think negative array indexing other than to append is an edge case, and would also be slightly confusing as it would differ from `list.insert()` where -1, -2, -3 inserts counting from just before the end instead of after, (hence I can't just delegate to `insert()`)
I think we can be fairly confident most common use cases are going to be append (default) and prepend, and positive-index as maybe a distant 3rd place. When writing tests you don't want to be worrying about keeping track of indexes in fixture code. I am going to make it raise an exception rather than leave as undefined behaviour.

Reply:
I would expect this to behave exactly like the `list.insert` as that is easy to document and explain ("it does the normal thing") rather than trying to document that it does something different (I had an off-by-one bug in my previous comments).

Reply:
but then you can't append to the handlers without keeping track of the indices, which makes it much less useful.

Reply:
I am surprised that `insert` can not put something last!  Could probably do that by taking `int | 'end'` and switch to `append` for 'end'.